File: src/main/java/com/somefrills/features/farming/autofarmer/MushroomMovement.java
package com.somefrills.features.farming.autofarmer;

/**
 * Mushroom pattern farming - always moving forward with lateral adjustments.
 * Cycles: Forward+Right → Forward only → Forward+Left → repeat
 */
public class MushroomMovement implements MovementStrategy {
    private State currentState = State.FORWARD_RIGHT;

    public MushroomMovement() {
    }

    public MovementState getCurrentState() {
        return currentState.toMovementState();
    }

    public void nextState() {
        currentState = currentState.next();
    }

    private enum State {
        FORWARD_RIGHT {
            @Override
            public State next() {
                return FORWARD_ONLY;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(true, false, false, true);
            }
        },
        FORWARD_ONLY {
            @Override
            public State next() {
                return FORWARD_LEFT;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(true, false, false, false);
            }
        },
        FORWARD_LEFT {
            @Override
            public State next() {
                return FORWARD_RIGHT;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(true, false, true, false);
            }
        };

        protected abstract State next();

        protected abstract MovementState toMovementState();
    }
}

File: src/main/kotlin/com/somefrills/features/farming/autofarmer/MushroomMovement.kt
package com.somefrills.features.farming.autofarmer

/**
 * Mushroom pattern farming - always moving forward with lateral adjustments.
 * Cycles: Forward+Right → Forward only → Forward+Left → repeat
 */
class MushroomMovement : MovementStrategy() {

    private var state: FarmingState = FarmingState.FORWARD_ONLY

    override fun getState(): MovementState {
        return state.toMovementState()
    }

    override fun nextState() {
        state = state.next() as FarmingState
    }

    private enum class FarmingState : State {
        FORWARD_ONLY {
            override fun next() = FORWARD_LEFT

            override fun toMovementState(): MovementState {
                return MovementState(MovementState.FORWARD)
            }
        },

        FORWARD_LEFT {
            override fun next() = FORWARD_ONLY

            override fun toMovementState(): MovementState {
                return MovementState(
                    MovementState.FORWARD or MovementState.LEFT or MovementState.SPRINT
                )
            }
        }
    }
}